lib/authorization.rb

```diff
@@ -18,7 +18,7 @@ def self.authorize(subject_id, actions, object_id)
   end
 
   def self.all_permissions(subject_id, object_id)
-    matched_policies(subject_id, nil, object_id).flat_map { _1[:actions] }
+    matched_policies_dataset(subject_id, nil, object_id).select_map(:actions).tap(&:flatten!)
```

Comment:
I didn't understand why `.tap(&:flatten!)` is used instead of simply `.flatten`

Reply:
`flatten` allocates an array, while `flatten!` modifies the receiver.  However, `flatten!` returns nil if there are no changes, so you can combine with to get the `flatten` behavior of always returning the array, but without the allocation.